tensorflow_addons/layers/crf.py

```diff
@@ -0,0 +1,478 @@
+            first_mask = left_boundary_mask[:, 0]
+
+        # remember this value for later use
+        self.mask = mask
+
+        if first_mask is not None:
+            no_left_padding = tf.math.reduce_all(first_mask)
+            msg = "Currently, CRF layer do not support left padding"
+            with tf.control_dependencies([
+                    tf.debugging.assert_equal(
+                        no_left_padding, tf.constant(True), message=msg)
+            ]):
+                self.potentials = self._dense_layer(inputs)
+        else:
+            self.potentials = self._dense_layer(inputs)
+
+        # appending boundary probability info
+        if self.use_boundary:
+            self.potentials = self.add_boundary_energy(
+                self.potentials, mask, self.left_boundary, self.right_boundary)
+
+        self.sequence_length = self._get_sequence_length(inputs, mask)
+
+        decoded_sequence, _ = self.get_viterbi_decoding(
+            self.potentials, self.sequence_length)
+
+        return decoded_sequence
+
+    def _get_sequence_length(self, input_, mask):
+        """Currently underline CRF fucntion (provided by
+        tensorflow_addons.text.crf) do not support bi-direction masking (left
+        padding / right padding), it support right padding by tell it the
+        sequence length.
+
+        this function is compute the sequence length from input and
+        mask.
+        """
+        if mask is not None:
+            sequence_length = self.mask_to_sequence_length(mask)
+        else:
+            # make a mask tensor from input, then used to generate sequence_length
+            input_energy_shape = tf.shape(input_)
+            raw_input_shape = tf.slice(input_energy_shape, [0], [2])
+            alt_mask = tf.ones(raw_input_shape)
+
+            sequence_length = self.mask_to_sequence_length(alt_mask)
+
+        return sequence_length
+
+    def mask_to_sequence_length(self, mask):
+        """compute sequence length from mask."""
+        sequence_length = tf.keras.backend.cast(
+            tf.keras.backend.sum(tf.keras.backend.cast(mask, tf.int8), 1),
+            tf.int64)
+        return sequence_length
+
+    @staticmethod
+    def _compute_mask_right_boundary(mask):
+        """input mask: 0011100, output left_boundary: 0000100."""
+        # shift mask to left by 1: 0011100 => 0111000
+        offset = 1
+        left_shifted_mask = tf.keras.backend.concatenate(
+            [mask[:, offset:],
+             tf.keras.backend.zeros_like(mask[:, :offset])],
+            axis=1)
+
+        # TODO(howl-anderson): for below code
+        # Original code in keras_contrib:
+        # end_mask = K.cast(
+        #   K.greater(self.shift_left(mask), mask),
+        #   K.floatx()
+        # )
+        # May have a bug, it's better confirmed
+        # by the original keras_contrib maintainer
+        # Luiz Felix (github: lzfelix),
+        # mailed him already and waiting for reply.
```

Comment:
Any update on this?

cc: @lzfelix

Reply:
Hello. My apologies in the delay, I was vacationing for a few days.

When this code was hosted under `keras-contrib` I fixed some issues regarding model saving/loading and at that time people asked me to be the maintainer of this layer, however I didn't write the code for it (and particularly find it really hard to follow).

Indeed, the CRF layer seems to have some bugs, for instance, loss becomes negative if the model is trained for too many iterations (notice that mathematically the CRF loss can't be negative), and this behaviour disappears when a masking layer is used.

Some users suggested potential bugs in the `keras-contrib` repo. I'll find all of them and update this comment, but it seems that this is one of them.

Reply:
Hi lzfelix, nice to have your reply. Since the bug with negative loss likely exists in this implementation too, I will try to find out why and fix it. I hope you can be the reviewer for those bugs (including this boundary one), that will be very helpful!

Reply:
Sure. I'll brush up my math on the CRF and help you with that!

Reply:
The bug which simillar to keras-team/keras-contrib#508 was already fixed in this PR.  This CRF layer uses `tensorflow_addons.text.crf` to compute the loss which is mainly ported from `tf.contrb.crf`, currently, as far as I know, there is no report on GitHub that `tf.contrb.crf` has a similar negative loss issue same with keras-team/keras-contrib#253, so far so good.